File: mobile/android/app/src/main/java/app/terra/explore/ModeSpeedGuard.kt
package app.terra.explore
class ModeSpeedGuard {
    private var highSince: Long?=null
    private var safeSince: Long?=null
    private var lastTime=0L
    var allowsDiscovery=true; private set
    var warning=false; private set
    fun update(kmh: Double,time: Long,mode: Mode) {
        if(!kmh.isFinite() || kmh<0 || time<=lastTime) return
        val limit=when(mode.category) { Mode.walk -> 22.0; Mode.bike -> 65.0; else -> Double.POSITIVE_INFINITY }
        if(time-lastTime>5000) { highSince=null; safeSince=null }; lastTime=time
        if(kmh>limit) { safeSince=null; if(highSince==null) highSince=time; allowsDiscovery=false; warning=time-(highSince ?: time)>=8000 }
        else { highSince=null; if(safeSince==null) safeSince=time; if(allowsDiscovery || time-(safeSince ?: time)>=3000) { allowsDiscovery=true; warning=false } }
    }
}

File: mobile/ios/Terra/ModeSpeedGuard.swift
import Foundation
struct ModeSpeedGuard {
    private var highSince: Double?
    private var lastTime: Double = 0
    private var safeSince: Double?
    private(set) var allowsDiscovery = true
    private(set) var warning = false
    mutating func update(kmh: Double,time: Double,mode: TravelMode) {
        guard kmh.isFinite,kmh >= 0,time > lastTime else { return }
        let limit: Double = mode.category == .walk ? 22 : mode.category == .bike ? 65 : .infinity
        if time-lastTime > 5000 { highSince = nil; safeSince = nil }; lastTime = time
        if kmh > limit { safeSince = nil; if highSince == nil { highSince = time }; allowsDiscovery = false; warning = time-(highSince ?? time) >= 8000 }
        else { highSince = nil; if safeSince == nil { safeSince = time }; if allowsDiscovery || time-(safeSince ?? time) >= 3000 { allowsDiscovery = true; warning = false } }
    }
}
